### PR TITLE
fix for #8664: Emit folding ranges for multi-line where clauses

### PR DESCRIPTION
**Pull request by m5tfi**
#8664 

I added a test that assert folding multi-line where clauses while leaving single lined one. Please, let me know if the code needs further improvements.

### CODE REVIEW

File: crates/ide/src/folding_ranges.rs [hunk 1]
```diff
@@ -109,6 +111,15 @@ pub(crate) fn folding_ranges(file: &SourceFile) -> Vec<Fold> {
...
                 }
+
+                // Fold where clause
+                if node.kind() == WHERE_CLAUSE && !visited_where_clauses.contains(&node) {
+                    if let Some(range) =
+                        contiguous_range_for_where(&node, &mut visited_where_clauses)
+                    {
+                        res.push(Fold { range, kind: FoldKind::WhereClause })
+                    }
+                }
```
Comment (matklad) [inline crates/ide/src/folding_ranges.rs:122]: These cases are for consequtive items of the same type, a where clause has an encompassing node, so this logic can be removed and instead a single new case added to `fold_kind`

Comment (m5tfi) [reply crates/ide/src/folding_ranges.rs:122]: But when we fold the where clause as a whole, it hides the `where` keyword

```rust
fn foo()
where
    A: Foo,
    B: Foo,
    C: Foo,
    D: Foo, {}
```
will be:

```rust
fn foo()
/* */ {}
```

instead of 

```rust
fn foo()
where /* */ {}
```
and i won't be able to check if it's a multi-line where clause or not, am I missing something ?

**Comment by matklad** [comment]
There is_multiline check being done above already. For not hiding the where
clause itself, yeah, I guess we need a special case, but it seems that the
special case is better fit for the common code-path, not for the
continious_range one.

**Comment by m5tfi** [comment]
> There is_multiline check being done above already

According to rustfmt ([here](https://rust-lang.github.io/rustfmt/?version=v1.4.36&search=#Where\%20predicates)). The `indent_style` of the where predicates defaults to "block". so, there will always be a newline `\n`.

```rust
fn foo()
where
    A: Foo,
{
}
```
Therefore, the `is_multiline` check above will always count where clauses as multi-line even if there was only one predicate.

So, I check if the first predicate matches the last one to make sure it's more than one, and start the fold at the end of the `where` keyword (second token) and end the fold at the end of the last comma (last token).

```rust
fn contiguous_range_for_where(
    node: &SyntaxNode,
    visited: &mut FxHashSet<SyntaxNode>,
) -> Option<TextRange> {
    let first_where_pred = node.first_child();
    let last_where_pred = node.last_child();

    if first_where_pred != last_where_pred {
        let mut it = node.descendants_with_tokens();
        if let (Some(_where_clause), Some(where_kw), Some(last_comma)) =
            (it.next(), it.next(), it.last())
        {
            let start = where_kw.text_range().end();
            let end = last_comma.text_range().end();

            visited.insert(node.clone());
            return Some(TextRange::new(start, end));
        }
    }
    None
}
```

I think my naming for the function might be misleading.

**Comment by m5tfi** [comment]
I am sorry, It's my third contribution ever on github and I made a couple of mistakes.

**TLDR** It is passing all the tests on my machine now.

I was working inside the `ide` crate and forgot to run tests for the whole project (outside the `ide` crate) before pushing. Then, I thought that I should merge with the recent commit in the original repo before trying to fix the failed test, but that failed another test (`tidy::check_merge_commits`) as I didn't know that I am not allowed to merge. So, I followed the stdout instructions, made a soft reset and fixed the issues.

Now, It's passing all of the tests on my machine. Can we try the github tests again please?

**Comment by matklad** [comment]
No worries about git! It took we a while to figure out a good workflow again. The tests are re-run automatically after every push.


> I think my naming for the function might be misleading.

Yeah, seems like it! It looks like we handle  where clauses like a special case -- they are unlike items like mods that can be folded completely, and they are unlike items like uses or commets, which are folded if there's a consequiteve group of them. 

So let's reflect this in code, and handle where clauses separaetly, without trying to mimic existing code. 

In  particular, `visited_where_clauses.contains(&node)` seems to be just a remenant of consequitve logic and isn't needed for where clauses?

**Comment by m5tfi** [comment]
> visited_where_clauses.contains(&node) seems to be just a remenant of consequitve logic and isn't needed for where clauses

That's true, sorry about that, i removed it now. I also renamed the function name to a better name (i think).

> The tests are re-run automatically after every push.

It keeps giving me this message : 
waiing

*Shame on you for not trusting noobs* (I am just kidding).

**Comment by m5tfi** [comment]
There's something I faced and couldn't solve. All the tests ran successfully locally on my Linux and on Github, but when I tried to run the tests locally on Windows10, It failed 2 tests:

* tests::lexer_tests
* tests::parser_tests

I checkout the oldest commit, before any of my changes, and these 2 tests still failed. Should I create an issue for it, or is it just something wrong with my development environment ? 

<details>

    failures:

    ---- tests::lexer_tests stdout ----


    error: expect test failed
       --> D:\monkey\rust-analyzer\crates/syntax/test_data\lexer/ok\0006_chars.txt

    You can update all `expect![[]]` tests by running:

        env UPDATE_EXPECT=1 cargo test

    To update a single test, place the cursor on `expect` token and use `run` feature of rust-analyzer.

    Expect:
    ----
    CHAR 3 "\'x\'"
    WHITESPACE 1 " "
    CHAR 3 "\' \'"
    WHITESPACE 1 " "
    CHAR 3 "\'0\'"
    WHITESPACE 1 " "
    CHAR 7 "\'hello\'"
    WHITESPACE 1 " "
    CHAR 6 "\'\\x7f\'"
    WHITESPACE 1 " "
    CHAR 4 "\'\\n\'"
    WHITESPACE 1 " "
    CHAR 4 "\'\\\\\'"
    WHITESPACE 1 " "
    CHAR 4 "\'\\\'\'"
    WHITESPACE 1 "\n"

    ----

    Actual:
    ----
    CHAR 3 "'x'"
    WHITESPACE 1 " "
    CHAR 3 "' '"
    WHITESPACE 1 " "
    CHAR 3 "'0'"
    WHITESPACE 1 " "
    CHAR 7 "'hello'"
    WHITESPACE 1 " "
    CHAR 6 "'\\x7f'"
    WHITESPACE 1 " "
    CHAR 4 "'\\n'"
    WHITESPACE 1 " "
    CHAR 4 "'\\\\'"
    WHITESPACE 1 " "
    CHAR 4 "'\\''"
    WHITESPACE 1 "\n"

    ----

    Diff:
    ----
    CHAR 3 "\'x\'"
    WHITESPACE 1 " "
    CHAR 3 "\' \'"
    WHITESPACE 1 " "
    CHAR 3 "\'0\'"
    WHITESPACE 1 " "
    CHAR 7 "\'hello\'"
    WHITESPACE 1 " "
    CHAR 6 "\'\\x7f\'"
    WHITESPACE 1 " "
    CHAR 4 "\'\\n\'"
    WHITESPACE 1 " "
    CHAR 4 "\'\\\\\'"
    WHITESPACE 1 " "
    CHAR 4 "\'\\\'\''"
    WHITESPACE 1 "\n"

    ----


    ---- tests::parser_tests stdout ----


    error: expect test failed
       --> D:\monkey\rust-analyzer\crates/syntax/test_data\parser/inline/ok\0003_where_pred_for.rast

    Expect:
    ----
    SOURCE_FILE@0..54
      FN@0..53
        FN_KW@0..2 "fn"
        WHITESPACE@2..3 " "
        NAME@3..12
          IDENT@3..12 "for_trait"
        GENERIC_PARAM_LIST@12..15
          L_ANGLE@12..13 "<"
          TYPE_PARAM@13..14
            NAME@13..14
              IDENT@13..14 "F"
          R_ANGLE@14..15 ">"
        PARAM_LIST@15..17
          L_PAREN@15..16 "("
          R_PAREN@16..17 ")"
        WHITESPACE@17..18 "\n"
        WHERE_CLAUSE@18..49
          WHERE_KW@18..23 "where"
          WHITESPACE@23..27 "\n   "
          WHERE_PRED@27..49
            FOR_KW@27..30 "for"
            GENERIC_PARAM_LIST@30..34
              L_ANGLE@30..31 "<"
              LIFETIME_PARAM@31..33
                LIFETIME@31..33
                  LIFETIME_IDENT@31..33 "\'a"
              R_ANGLE@33..34 ">"
            WHITESPACE@34..35 " "
            PATH_TYPE@35..36
              PATH@35..36
                PATH_SEGMENT@35..36
                  NAME_REF@35..36
                    IDENT@35..36 "F"
            COLON@36..37 ":"
            WHITESPACE@37..38 " "
            TYPE_BOUND_LIST@38..49
              TYPE_BOUND@38..49
                PATH_TYPE@38..49
                  PATH@38..49
                    PATH_SEGMENT@38..49
                      NAME_REF@38..40
                        IDENT@38..40 "Fn"
                      PARAM_LIST@40..49
                        L_PAREN@40..41 "("
                        PARAM@41..48
                          REF_TYPE@41..48
                            AMP@41..42 "&"
                            LIFETIME@42..44
                              LIFETIME_IDENT@42..44 "\'a"
                            WHITESPACE@44..45 " "
                            PATH_TYPE@45..48
                              PATH@45..48
                                PATH_SEGMENT@45..48
                                  NAME_REF@45..48
                                    IDENT@45..48 "str"
                        R_PAREN@48..49 ")"
        WHITESPACE@49..50 "\n"
        BLOCK_EXPR@50..53
          L_CURLY@50..51 "{"
          WHITESPACE@51..52 " "
          R_CURLY@52..53 "}"
      WHITESPACE@53..54 "\n"

    ----

    Actual:
    ----
    SOURCE_FILE@0..54
      FN@0..53
        FN_KW@0..2 "fn"
        WHITESPACE@2..3 " "
        NAME@3..12
          IDENT@3..12 "for_trait"
        GENERIC_PARAM_LIST@12..15
          L_ANGLE@12..13 "<"
          TYPE_PARAM@13..14
            NAME@13..14
              IDENT@13..14 "F"
          R_ANGLE@14..15 ">"
        PARAM_LIST@15..17
          L_PAREN@15..16 "("
          R_PAREN@16..17 ")"
        WHITESPACE@17..18 "\n"
        WHERE_CLAUSE@18..49
          WHERE_KW@18..23 "where"
          WHITESPACE@23..27 "\n   "
          WHERE_PRED@27..49
            FOR_KW@27..30 "for"
            GENERIC_PARAM_LIST@30..34
              L_ANGLE@30..31 "<"
              LIFETIME_PARAM@31..33
                LIFETIME@31..33
                  LIFETIME_IDENT@31..33 "'a"
              R_ANGLE@33..34 ">"
            WHITESPACE@34..35 " "
            PATH_TYPE@35..36
              PATH@35..36
                PATH_SEGMENT@35..36
                  NAME_REF@35..36
                    IDENT@35..36 "F"
            COLON@36..37 ":"
            WHITESPACE@37..38 " "
            TYPE_BOUND_LIST@38..49
              TYPE_BOUND@38..49
                PATH_TYPE@38..49
                  PATH@38..49
                    PATH_SEGMENT@38..49
                      NAME_REF@38..40
                        IDENT@38..40 "Fn"
                      PARAM_LIST@40..49
                        L_PAREN@40..41 "("
                        PARAM@41..48
                          REF_TYPE@41..48
                            AMP@41..42 "&"
                            LIFETIME@42..44
                              LIFETIME_IDENT@42..44 "'a"
                            WHITESPACE@44..45 " "
                            PATH_TYPE@45..48
                              PATH@45..48
                                PATH_SEGMENT@45..48
                                  NAME_REF@45..48
                                    IDENT@45..48 "str"
                        R_PAREN@48..49 ")"
        WHITESPACE@49..50 "\n"
        BLOCK_EXPR@50..53
          L_CURLY@50..51 "{"
          WHITESPACE@51..52 " "
          R_CURLY@52..53 "}"
      WHITESPACE@53..54 "\n"

    ----

    Diff:
    ----
    SOURCE_FILE@0..54
      FN@0..53
        FN_KW@0..2 "fn"
        WHITESPACE@2..3 " "
        NAME@3..12
          IDENT@3..12 "for_trait"
        GENERIC_PARAM_LIST@12..15
          L_ANGLE@12..13 "<"
          TYPE_PARAM@13..14
            NAME@13..14
              IDENT@13..14 "F"
          R_ANGLE@14..15 ">"
        PARAM_LIST@15..17
          L_PAREN@15..16 "("
          R_PAREN@16..17 ")"
        WHITESPACE@17..18 "\n"
        WHERE_CLAUSE@18..49
          WHERE_KW@18..23 "where"
          WHITESPACE@23..27 "\n   "
          WHERE_PRED@27..49
            FOR_KW@27..30 "for"
            GENERIC_PARAM_LIST@30..34
              L_ANGLE@30..31 "<"
              LIFETIME_PARAM@31..33
                LIFETIME@31..33
                  LIFETIME_IDENT@31..33 "\'a"
              R_ANGLE@33..34 ">"
            WHITESPACE@34..35 " "
            PATH_TYPE@35..36
              PATH@35..36
                PATH_SEGMENT@35..36
                  NAME_REF@35..36
                    IDENT@35..36 "F"
            COLON@36..37 ":"
            WHITESPACE@37..38 " "
            TYPE_BOUND_LIST@38..49
              TYPE_BOUND@38..49
                PATH_TYPE@38..49
                  PATH@38..49
                    PATH_SEGMENT@38..49
                      NAME_REF@38..40
                        IDENT@38..40 "Fn"
                      PARAM_LIST@40..49
                        L_PAREN@40..41 "("
                        PARAM@41..48
                          REF_TYPE@41..48
                            AMP@41..42 "&"
                            LIFETIME@42..44
                              LIFETIME_IDENT@42..44 "\'a"
                            WHITESPACE@44..45 " "
                            PATH_TYPE@45..48
                              PATH@45..48
                                PATH_SEGMENT@45..48
                                  NAME_REF@45..48
                                    IDENT@45..48 "str"
                        R_PAREN@48..49 ")"
        WHITESPACE@49..50 "\n"
        BLOCK_EXPR@50..53
          L_CURLY@50..51 "{"
          WHITESPACE@51..52 " "
          R_CURLY@52..53 "}"
      WHITESPACE@53..54 "\n"

    ----



    failures:
        tests::lexer_tests
        tests::parser_tests

</details>

**Comment by Veykril** [comment]
The debug output has changed in recent rust nightlies which causes that failure, I imagine your windows setup uses a recent nightly toolchain while your linux setup isnt

**Comment by lnicola** [comment]
That's #8585.

**Comment by matklad** [comment]
bors r+

looks good to me, thanks!
